### PR TITLE
plugin: implement collapsible code for protosaurus language using metastring

### PR DESCRIPTION
**Pull request by imballinst**
This PR implements the collapsible code for protosaurus language, utilizing the metastring. Key features:

1. Code title (with `title="..."`)
2. Collapsible code with (`collapsible`)

Link to page: https://imballinst.github.io/protosaurus/pr-33/docs/test

image

image

image

## How it works

1. We parse the metastring
2. If there is only title, then we add the title element before the code
4. If there is only collapsible, throw an error (included in test)
5. If there are both, then the title is used as the collapsible title

## Next steps

1. Apply it for non-protosaurus language as well (currently we need to use `<details>` tag
2. Implement line highlighting for protosaurus language using metastring
4. Do all the minor `TODO`s.
5. Perhaps it's a good time to refactor `packages/rehype-plugin-codeblock/src/index.ts`.

Signed-off-by: Try Ajitiono <ballinst@gmail.com>

### CODE REVIEW

File: packages/rehype-plugin-codeblock/src/hast.ts [hunk 1]
```diff
@@ -72,6 +73,60 @@ export function getHastElementType(
...
+                  value: metastringInfo.title
+                }
+              ]
+            }
+          ]
+        },
+        element
+      ]
+    });
+  } else if (metastringInfo.title) {
```
Comment (dio) [inline packages/rehype-plugin-codeblock/src/hast.ts]: Optional, add description on the structure that we push into children.

**Review by dio — APPROVED** [review]

File: packages/rehype-plugin-codeblock/src/hast.ts (hunk 1 above)
Comment (imballinst) [reply packages/rehype-plugin-codeblock/src/hast.ts]: Ah, yes. Sure! I can do that tomorrow. Thanks for the suggestion!

Comment (imballinst) [reply packages/rehype-plugin-codeblock/src/hast.ts]: @dio I added comments which represent the resulting JSX structure:

https://github.com/imballinst/protosaurus/blob/9fdc7572108501364c8d2e3e8e90425bcd8f3996/packages/rehype-plugin-codeblock/src/hast.ts#L84-L111

Let me know what you think! I'll merge this for now. If there are more things that you'd like to happen, I'll add it to my to-do.

**Comment by imballinst** [comment]
Thanks @dio for the review!

**Pull request merged**